File: 1.TwoSum.ts
// Two Sum - Brute Force Approach
// Time Complexity: O(n²) - nested loops iterate through array
// Space Complexity: O(1) - only using constant extra space
function twoSum(nums: number[], target: number): number[] {
    // Outer loop: iterate through each element (n iterations)
    for(let i=0;i<nums.length;i++){
     // Inner loop: check remaining elements (n-1, n-2, ... iterations)
     // This creates O(n²) time complexity due to nested iteration
     for(let j=i+1;j<nums.length;j++){
          // Check if current pair sums to target
          if (nums[i] + nums[j] === target){
                return[i,j] // Return indices of the two numbers
          }
     }
    }
    return [] // Return empty array if no solution found
};

console.log(twoSum([2,5,5,11],10))

File: 1.TwoSum.py
# Author: madhu
# Two Sum problem solution using hash map approach

def two_sum(nums, target):
    """
    Find two numbers in the array that add up to the target sum.
    Returns the indices of the two numbers.
    """
    seen = {}  # Hash map to store number -> index mapping
    for i, num in enumerate(nums):
        needed = target - num  # Calculate what number we need to find
        if needed in seen:  # Check if the needed number exists in our hash map
            return [seen[needed], i]  # Return indices of the two numbers
        seen[num] = i  # Store current number and its index
    return []  # Return empty array if no solution found


# Test the function
num = [2,7,11,13]
target = 20
result = two_sum(num,target)
print(f"result is {result}")